jinahub/indexers/cache/__init__.py

```diff
@@ -0,0 +1,139 @@
+            self.hash_to_id = dict()
+
+    def close(self):
+        """Flushes the in-memory cache to pickle files."""
+        pickle.dump(self.id_to_hash, open(self.path + '.ids', 'wb'))
+        pickle.dump(self.hash_to_id, open(self.path + '.cache', 'wb'))
+
+
+default_fields = (CONTENT_HASH_KEY,)
+
+
+class DocCache(Executor):
+    """An indexer that caches combinations of fields
+    and filters out documents that have been previously cached"""
+
+    def __init__(
+        self,
+        fields: Optional[Tuple[str]] = None,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        if fields is None:
+            fields = default_fields
+        self.fields = fields
+        self.logger = get_logger(self)
+        os.makedirs(self.workspace)
+        self.cache_handler = _CacheHandler(
+            os.path.join(self.workspace, 'cache'), self.logger
+        )
+
+    @requests(on='/index')
+    def index(self, docs: DocumentArray, **kwargs):
+        """Index Documents in the cache, by hashing self.fields
+
+        If the document was already previously cached,
+        it is removed from the docs, so no further Executor will receive it
+
+        :param docs: the documents to cache"""
+
+        indices_to_remove = []
+        for i, d in enumerate(docs):
+            doc_hash = DocCache.hash_doc(d, self.fields)
+            exists = doc_hash in self.cache_handler.hash_to_id.keys()
+
+            if not exists:
+                self.cache_handler.id_to_hash[d.id] = doc_hash
+                self.cache_handler.hash_to_id[doc_hash] = d.id
+            else:
+                indices_to_remove.append(i)
+
+        indices_to_remove = sorted(indices_to_remove, reverse=True)
+        for i in indices_to_remove:
+            del docs[i]
+
+    def close(self) -> None:
+        self.cache_handler.close()
+
+    @staticmethod
+    def hash_doc(doc: Document, fields: Tuple[str]) -> bytes:
+        """Calculate hash by which we cache.
+
+        :param doc: the Document
+        :param fields: the list of fields
+        :return: the hash value of the fields
+        """
+        values = doc.get_attributes(*fields)
+        if not isinstance(values, list):
+            values = [values]
+        data = ''
+        for field, value in zip(fields, values):
+            data += f'{field}:{value};'
+        digest = hashlib.sha256(bytes(data.encode('utf8'))).digest()
+        return digest
+
+    @property
+    def size(self):
+        assert len(self.cache_handler.id_to_hash) == len(
+            self.cache_handler.hash_to_id
+        ), 'Cache state invalid. Length of mappings does not align. Try rebuilding the index'
+        return len(self.cache_handler.id_to_hash)
+
+    @requests(on='/update')
+    def update(self, docs: DocumentArray, **kwargs):
+        """Update the documents in the cache with the new content
+
+        WARNING: If the Document doesn't exist in the cache
```

Comment:
Feels like this sentence is not complete.

Reply:
removed